Userland/Libraries/LibJS/Runtime/Value.cpp

```diff
@@ -517,6 +500,96 @@ ThrowCompletionOr<Value> Value::to_number(VM& vm) const
     }
 }
 
+constexpr bool is_ascii_number(u32 code_point)
+{
+    return is_ascii_digit(code_point) || code_point == '.' || (code_point == 'e' || code_point == 'E') || code_point == '+' || code_point == '-';
+}
+
+struct NumberParseResult {
+    StringView literal;
+    u8 base;
+};
+
+static Optional<NumberParseResult> parse_number_text(StringView text)
+{
+    NumberParseResult result {};
+
+    auto check_prefix = [&](auto lower_prefix, auto upper_prefix) {
+        if (text.length() <= 2)
+            return false;
+        if (!text.starts_with(lower_prefix) && !text.starts_with(upper_prefix))
+            return false;
+        return true;
+    };
+
+    // https://tc39.es/ecma262/#sec-tonumber-applied-to-the-string-type
+    if (check_prefix("0b"sv, "0B"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_binary_digit))
+            return {};
+
+        result.literal = text.substring_view(2);
+        result.base = 2;
+    } else if (check_prefix("0o"sv, "0O"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_octal_digit))
+            return {};
+
+        result.literal = text.substring_view(2);
+        result.base = 8;
+    } else if (check_prefix("0x"sv, "0X"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_hex_digit))
+            return {};
+
+        result.literal = text.substring_view(2);
+        result.base = 16;
+    } else {
+        if (!all_of(text, is_ascii_number))
+            return {};
+
+        result.literal = text;
+        result.base = 10;
+    }
+
+    return result;
+}
+
+// 7.1.4.1.1 StringToNumber ( str ), https://tc39.es/ecma262/#sec-stringtonumber
+Optional<Value> Value::string_to_number() const
+{
+    VERIFY(is_string());
+
+    // 1. Let text be StringToCodePoints(str).
+    String string = Utf8View(as_string().string()).trim(whitespace_characters, AK::TrimMode::Both).as_string();
+
+    // 2. Let literal be ParseText(text, StringNumericLiteral).
+    if (string.is_empty())
+        return Value(0);
+    if (string == "Infinity" || string == "+Infinity")
+        return js_infinity();
+    if (string == "-Infinity")
+        return js_negative_infinity();
+
+    auto result = parse_number_text(string);
+
+    // 3. If literal is a List of errors, return NaN.
+    if (!result.has_value())
+        return js_nan();
+
+    if (result->base != 10) {
+        auto bigint = Crypto::UnsignedBigInteger::from_base(result->base, result->literal);
+
+        // 4. Return StringNumericValue of literal.
+        return Value(bigint.to_double());
+    }
+
+    char* endptr;
+    auto parsed_double = strtod(string.characters(), &endptr);
+    if (*endptr)
+        return js_nan();
+
+    // 4. Return StringNumericValue of literal.
+    return Value(parsed_double);
```

Comment:
Please don't repeat spec steps

```suggestion
    // 4. Return StringNumericValue of literal.
    if (result->base != 10) {
        auto bigint = Crypto::UnsignedBigInteger::from_base(result->base, result->literal);
        return Value(bigint.to_double());
    }

    char* endptr;
    auto parsed_double = strtod(string.characters(), &endptr);
    if (*endptr)
        return js_nan();

    return Value(parsed_double);
```